hugr-core/src/types/signature.rs

```diff
@@ -118,6 +120,28 @@ impl<RV: MaybeRV> FuncTypeBase<RV> {
         self.output.validate(extension_registry, var_decls)?;
         self.extension_reqs.validate(var_decls)
     }
+
+    /// Returns a registry with the concrete extensions used by this signature.
+    ///
+    /// Note that extension type parameters are not included, as they have not
+    /// been instantiated yet.
+    ///
+    /// This method only returns extensions actually used by the types in the
+    /// signature. The extension deltas added via [`Self::with_extension_delta`]
+    /// refer to _runtime_ extensions, which may not be in all places that
+    /// manipulate a HUGR.
+    pub fn used_extensions(&self) -> Result<ExtensionRegistry, ExtensionCollectionError> {
```

Comment:
So the other way to do this - is to put the `collect_blah_exts` methods as impls of a single method in a trait, and then you can add a trait-default method (or a method taking an `impl ExtensionCollectionTrait`) that does this...that'd get you the two `used_extensions` methods that you have plus a load more with only having to write this once. (I wasn't going to suggest the trait approach until I saw this...)

Reply:
This doesn't work if you change `OpType::used_extensions` to take the extra `Option<Node>` parameter I suggested elsewhere, though. So feel free to drop.

Reply:
I opted for keeping all the logic in a single module, as it's short and quite similar between all the ops.